[VPlan][LV] Add removeRedundantAndMasks to VPlanTransforms

nasherm opened the pull request:
This patch adds the removeRedundantAndMasks to VPlanTransforms and including
  appropriate tests. It's been found when benchmarking the vectorizer
  and loop flattening that we sometimes generate loops of the following
  basic form

```
vector.scevcheck:                                 ; preds = %for.cond1.preheader.lr.ph
  %cmp = icmp ugt i64 %N, 4294967295
  br i1 %cmp, label %end, label %vector.body
vector.body:                                      ; preds = %vector.body, %vector.ph
  %index = phi i64 [ 0, %vector.scevcheck ], [ %index.next, %vector.body ]
  %and = and i64 %index, 4294967295
  %index.next = add i64 %and, 1
  %exit.cond = icmp ugt i64 %index.next, %N
  br i1 %exit.cond, label %end, label %vector.body
end:
  ret i64 %N
```

In this loop the 'and' mask is introduced as a bounds check
but is a no-op due to the fact the runtime bounds check block
'vector.scevcheck' telegraphs information about the maximum tripcount
of this loop. This patch allows for transformations that turn the above
loop to

```
vector.scevcheck:                                 ; preds = %for.cond1.preheader.lr.ph
  %cmp = icmp ugt i64 %N, 4294967295
  br i1 %cmp, label %end, label %vector.body
vector.body:                                      ; preds = %vector.body, %vector.ph
  %index = phi i64 [ 0, %vector.scevcheck ], [ %index.next, %vector.body ]
  %index.next = add i64 %index, 1
  %exit.cond = icmp ugt i64 %index.next, %N
  br i1 %exit.cond, label %end, label %vector.body
end:
  ret i64 %N
```

With this patch we've seen performance improvements of up to 8% on internal
benchmarks tested on AArch64 cores.

nasherm committed:
[Transforms] Add LoopNoOpElimination pass

This patch adds the LoopNoOpEliminatioin pass including
appropriate tests. It's been found when benchmarking the vectorizer
and loop flattening that we sometimes generate loops of the following
basic form

```
vector.scevcheck:                                 ; preds = %for.cond1.preheader.lr.ph
  %cmp = icmp ugt i64 %N, 4294967295
  br i1 %cmp, label %end, label %vector.body
vector.body:                                      ; preds = %vector.body, %vector.ph
  %index = phi i64 [ 0, %vector.scevcheck ], [ %index.next, %vector.body ]
  %and = and i64 %index, 4294967295
  %index.next = add i64 %and, 1
  %exit.cond = icmp ugt i64 %index.next, %N
  br i1 %exit.cond, label %end, label %vector.body
end:
  ret i64 %N
```

In this loop the 'and' mask is introduced as a bounds check
but is a no-op due to the fact the runtime bounds check block
'vector.scevcheck' telegraphs information about the maximum tripcount
of this loop. This patch allows for transformations that turn the above
loop to

```
vector.scevcheck:                                 ; preds = %for.cond1.preheader.lr.ph
  %cmp = icmp ugt i64 %N, 4294967295
  br i1 %cmp, label %end, label %vector.body
vector.body:                                      ; preds = %vector.body, %vector.ph
  %index = phi i64 [ 0, %vector.scevcheck ], [ %index.next, %vector.body ]
  %index.next = add i64 %index, 1
  %exit.cond = icmp ugt i64 %index.next, %N
  br i1 %exit.cond, label %end, label %vector.body
end:
  ret i64 %N
```

With this patch we've seen performance improvements of up to 8% on internal
benchmarks.

Change-Id: I9bb50138bf92da41a94173a5c6da9131e839560b

nasherm committed:
Enable loop-noop-elim in test

Enable loop-noop-elim in loop-no-op-and-elim.ll test
to highlight the function of the pass.

Change-Id: I2bd963f5390efc8c45e8caa5199729b2a547807a

fhahn reviewed (commented):
There's already loop-deletion, which should remove such dead loops?

fhahn reviewed (commented):
Ah presumably in the real case the whole loop isn't a no-op.

nasherm committed:
Code formatting

Change-Id: I004c6b1c3a6889e20d0138c59bd570eadfd7896e

nikic reviewed (commented):
This seems like something IndVarSimplify should handle.

nikic commented:
If you are specifically targeting post-vectorization here, is it possible to do this as a VPlan simplification instead?

nasherm committed:
Move LoopNoOpElim logic to VPlan

This patch responds to review comments by moving
LoopNoOpElimination logic to a VPlanTransform

Change-Id: I59651481c17595d9a1ec6c5e3b1bebef157378a2

nasherm committed:
Enable VPlanTransforms::removeRedundantAndMasks

Enable removeRedundantAndMasks in VPlanTransform LoopVectorization
transforms pipeline. This highlights the optimization and how it affects
relevant tests.

Change-Id: I8daa53646931be0d90bd93f1e2f937916e7332b1

nasherm committed:
Fix test crash

Change-Id: If89ee3a4c67ab4589f15481bdf99818bce7e6795

nasherm commented:
> If you are specifically targeting post-vectorization here, is it possible to do this as a VPlan simplification instead?

Yes! Sorry for the delay. It took me a while to get up to speed with VPlan. I've ported over my original pass logic into VPlanTransforms

nasherm committed:
Code formatting

Change-Id: I2da607e2366c641052134747bcef9475be10da26
